src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultRazorDocumentMappingService.cs

```diff
@@ -18,77 +18,156 @@ namespace Microsoft.AspNetCore.Razor.LanguageServer
+                // The source mapping starts at char 13 on the "@functions" line (after the open brace). Unfortunately
+                // and code that is needed on that line, say an attribute that the code action wants to insert, will
+                // start at char 8 because of the desired indentation of that new code. This means it starts outside of the
+                // mapping, so is thrown away, which results in data loss.
+                //
+                // To fix this we check and if the mapping would have been successful at the end of the line (char 13 above)
+                // then we insert a newline, and enough indentation to get us back out to where the new code wanted to start (char 8)
+                // and then we're good - we've left the @functions bit alone which razor needs, but we're still able to insert
+                // new code above where the C# code is in the generated document.
+                //
+                // One last hurdle is that sometimes these edits come in as separate edits. So for example replacing "class Goo" above
+                // with "public class Goo" would come in as one edit for "public", one for "class" and one for "Goo", all on the same line.
+                // When we map the edit for "public" we will push everything down a line, so we don't want to do it for other edits
+                // on that line.
+                if (!mappedStart && !mappedEnd && !range.SpansMultipleLines())
                 {
-                    return true;
+                    // If the new text doesn't have any content we don't care - throwing away invisible whitespace is fine
+                    if (string.IsNullOrWhiteSpace(edit.NewText))
+                    {
+                        continue;
+                    }
+
+                    var line = csharpSourceText.Lines[range.Start.Line];
+
+                    // If the line isn't blank, then this isn't a functions directive
+                    if (line.GetFirstNonWhitespaceOffset() is not null)
+                    {
+                        continue;
+                    }
+
+                    // Only do anything if the end of the line in question is a valid mapping point (ie, a transition)
+                    var endOfLine = line.Span.End;
+                    if (TryMapFromProjectedDocumentPosition(codeDocument, endOfLine, out var hostDocumentIndex, out _))
+                    {
+                        if (range.Start.Line == lastNewLineAddedToLine)
+                        {
+                            // If we already added a newline to this line, then we don't want to add another one, but
+                            // we do need to add a space between this edit and the previous one, because the normalization
+                            // will have swallowed it.
+                            projectedEdits.Add(new TextEdit()
+                            {
+                                NewText = " " + edit.NewText,
```

Comment:
This is the only bit I'm a little concerned about because it's trying to understand what types of edits are flowing into the system and that they could be piecemealed

Reply:
So the edits coming in to the system are straightforward (ish) and you can see them in the tests. What happens though is we call our `NormalizeEdits` method, which applies the edits then gets changes from a diff, and processes that. I _assume_ its done that way because mapping the start and end of an edit is not very useful when its a huge one, as its more likely to fail, so smaller edits definitely make sense.